### PR TITLE
Downscaling and bias-correction module

### PR DESCRIPTION
**Pull request by aulemahal**
### Pull Request Checklist:
- [ ] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #xyz
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Documentation has been added / updated (for bug fixes / features)
- [ ] HISTORY.rst has been updated (with summary of main changes)
- [ ] `bumpversion (minor / major / patch)` has been called on this branch
- [ ] Tags have been pushed (`git push --tags`)

* **What kind of change does this PR introduce?** 

Development of a new `xclim.downscaling` module for downscaling and post-processing of climate data.  Decision was made to implement our own version because we wanted to stay as close as possible to the xclim way-of-life, i.e.: 1) follow xarray and dask tightly, with a minimum of other dependencies, 2) a programming style that is easy to understand for science people already used to xarray.   The main alternative, `scikit-downscale` did not completely fit both conditions.

Moreover, after discussion with @Paattrriicckk, we wanted something that is not Method-specific, but rather modular. C-à-d: We don't want a "Cannon et al. (2015)" method, but rather a bias-correction pipeline that uses this kind of detrending and this other kind of mapping with these specific parameters. Thus, what I suggest here is as follows:

Each substep of the process has different flavors : Different subclasses of a main generic class.  The final processing is either done by provided "pipelines" or just by the user chaining appropriate instances 

Copy from the doc in xclim/downscaling/__init__.py:

> What the use calls is a "Pipeline".
> Given a detrending, a grouping and a mapping object, a typical pipeline would call:
> 
> - detrend.fit() + detrend.detrend() on each of obs, sim and fut
> - grouping.group() on obs and sim
> - mapping.fit() on obs+sim
> - grouping.add_group_axis() on fut
> - mapping.predict() on fut
> - detrend.retrend() on fut
> 
> The whole module assumes the bias-correction is made along coord "time", for each group in coord "group", independent of any other coordinates.
> 
> The Detrending objects will detrend/retrend DataArrays along "time".
> 
> The Grouping objects will group elements of a DataArray along time and give back either a DataArrayGroupBy or a DataArray object with a time coord on which mapping should be done.
> Once the action called, the new dimension is called "group" and has a "group_name" attribute with the real group name (the one a simple groupby(dim) would have given.)
> Also, the add_group_axis() method, adds a coord that says where is each element along the time coord in the group. Ex : If the grouping is monthly, the 15th of April  group = 4, the 30th group = 4.5.
> 
> The Mapping objects perform a fit from obs and sim, the reference observation and the simulated climate for the same reference period, the fit is performed point-wise and group-wise, along "time".
> Then, the predict() method takes in fut, the simulated climate for the projection period, that must have a "time" and a "group" coord, as given by the add_group_axis() grouping method.
> It returns the corrected fut timeseries.

Example from `pipeline.py`
```python
def basicpipeline(
    obs: DataArray,
    sim: DataArray,
    fut: DataArray,
    detrender=NoDetrend(),
    grouper=EmptyGrouping(),
    mapper=DeltaMapping(),
):
    obs_trend = detrender.fit(obs)  # <- Each call to fit() returns a new copy of the detrending object
    sim_trend = detrender.fit(sim)  #    that has __fitted == True.
    fut_trend = detrender.fit(fut)

    obs = obs_trend.detrend(obs)
    sim = sim_trend.detrend(sim)
    fut = fut_trend.detrend(fut)

    mapper.fit(grouper.group(obs), grouper.group(sim))  # <- compute fitting parameters
    fut_corr = mapper.predict(grouper.add_group_axis(fut))  # <- apply them

    fut_corr = fut_trend.retrend(fut_corr)

    return fut_corr
```

In the current structure, instances of detrending, grouping and mapping classes are initialized with their parameters and then used on data. Subclasses should only implement the private (with '_') versions of the methods so that the generic classes take care of basic checks and that the signature stays the same always.

This is all work-in-progress, what looks now simple might be not as maintainable as I think.... Let's discuss it here!

* **Does this PR introduce a breaking change?** 
No

### CODE REVIEW

File: xclim/downscaling/detrending.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+class NoDetrend(ParametrizableClass):
+    """Base class for detrending objects
+
+    Defines three methods:
+
+    fit(da)     : Compute trend from da and return a new _fitted_ Detrend object.
+    detrend(da) : Return detrended array.
+    retrend(da) : Puts trend back on da.
+
+    * Subclasses should implement _fit(), _detrend() and _retrend(), not the methods themselves.
```
[the remark this answers is not in the document]
Comment (aulemahal) [reply xclim/downscaling/detrending.py]: So we can perform some tests before calling them. Ex: detrend() raises an error if fit() hasn't been called. And _fit() doesn't have to include to class creation code.

File: xclim/downscaling/__init__.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+climate data using xarray. Once complete and usable, chances are that this
+submodule will move into its own package, so don't get attached too much.
+
+
+Logic:
+
+What the use calls is a "Pipeline".
+Given a detrending, a grouping and a mapping object, a typical pipeline would call:
+
+- detrend.fit() + detrend.detrend() on each of obs, sim and fut
```
[the remark this answers is not in the document]
Comment (aulemahal) [reply xclim/downscaling/__init__.py]: Hum, je ne savais pas! C'est sur que ça modifie la structure...

**Comment by aulemahal** [comment]
D'accord. Je voulais simplement éviter qu'on finisse avec un code monstrueux comme downscaleR ou comme tous les scripts matlab du monde.

Mettons la structure en quarantaine pour le moment.

**Comment by huard** [comment]
Je suis sensible à cet argument, et entièrement d'accord. Je pense qu'on va rapidement réaliser qu'on est capable de faire entrer des morceaux dans cette structure, et d'autres non. On avisera pour trouver un modèle flexible.

**Pull request merged**